programs/bpf_loader/src/syscalls/cpi.rs

```diff
@@ -879,73 +887,411 @@ fn cpi_common<S: SyscallInvokeSigned>(
-                return Err(
-                    SyscallError::InstructionError(InstructionError::AccountDataTooSmall).into(),
+        }
+
+        fn invoke_context(&mut self) -> InvokeContext<'_> {
+            let mut invoke_context = InvokeContext::new_mock(&mut self.transaction_context, &[]);
+            invoke_context
+                .transaction_context
+                .get_next_instruction_context()
+                .unwrap()
+                .configure(
+                    &self.program_accounts,
+                    &self.preparation.instruction_accounts,
+                    &self.instruction_data,
                 );
+            invoke_context.push().unwrap();
+            invoke_context
+        }
+    }
+
+    struct MockCallerAccount {
+        lamports: u64,
+        owner: Pubkey,
+        vm_addr: u64,
+        data: Vec<u8>,
+        len: u64,
+        region: MemoryRegion,
+    }
+
+    impl MockCallerAccount {
+        fn new(lamports: u64, owner: Pubkey, vm_addr: u64, data: &[u8]) -> MockCallerAccount {
+            // write [len][data] into a vec so we can check that they get
+            // properly updated by update_caller_account()
+            let mut d = vec![0; mem::size_of::<u64>() + data.len() + MAX_PERMITTED_DATA_INCREASE];
+            unsafe { ptr::write_unaligned::<u64>(d.as_mut_ptr().cast(), data.len() as u64) };
+            d[mem::size_of::<u64>()..][..data.len()].copy_from_slice(data);
+
+            // the memory region must include the realloc data
+            let region = MemoryRegion::new_writable(d.as_mut_slice(), vm_addr);
+
+            // caller_account.data must have the actual data length
+            d.truncate(mem::size_of::<u64>() + data.len());
+
+            MockCallerAccount {
+                lamports,
+                owner,
+                vm_addr,
+                data: d,
+                len: data.len() as u64,
+                region,
             }
-            to_slice.copy_from_slice(from_slice);
         }
+
+        fn data_slice<'a>(&self) -> &'a [u8] {
+            // lifetime crimes
+            unsafe {
+                slice::from_raw_parts(
+                    self.data[mem::size_of::<u64>()..].as_ptr(),
+                    self.data.capacity() - mem::size_of::<u64>(),
+                )
+            }
+        }
+
+        fn caller_account(&mut self) -> CallerAccount<'_> {
+            let data = &mut self.data[mem::size_of::<u64>()..];
+            CallerAccount {
+                lamports: &mut self.lamports,
+                owner: &mut self.owner,
+                original_data_len: data.len() as usize,
+                data,
+                vm_data_addr: self.vm_addr + mem::size_of::<u64>() as u64,
+                ref_to_len_in_vm: &mut self.len,
+                executable: false,
+                rent_epoch: 0,
+            }
+        }
+    }
+
+    fn one_instruction_account(data: Vec<u8>) -> Vec<TestTransactionAccount> {
```

Comment:
Confusing name, I guess what this method is ought to do is to create a "transaction with one instruction account".